packages/adminrouter/extra/src/README.md

```diff
@@ -264,28 +264,178 @@ it, but it does not seem to be justified. For example:
+* namespace and name: `foo/bar/baz/picture.jpg`, resource address: `/`
+
+So the algorithm iteratively tries to resolve each "candidate" namespace and name
+and assume that first-match wins.
+
+Assuming that we deployed `foo/bar` NGINX task, we will see two iterations:
+* first one with `foo` candidate namespace+service, which is not present in the
+cluster
+* second one with `foo/bar` candidate namespace+service name, which results in
+request being routed to correct backend. In case when user is trying to access
+an non-existent service, there are going to be four iterations, each one
+unsuccessful. After the fourth one the code will give up.
+
+There is of course a limit on the number of iterations that AR makes. Currently,
+it is set to `10`. This means that a service cannot nest more than nine times and
+that everything after the tenth segment is always treated as a resource component.
+
+#### Resolving individual component
+During each iteration described in previous paragraph, AR tries to resolve the
+component using the data stored in the cache and if necessary - from MesosDNS:
+* first step is to check if there is a root Marathon task that has lebels set
+  and these lebels:
+  * `DCOS_SERVICE_NAME` label contents is equal to given `namespace+service` string
+  * `DCOS_SERVICE_SCHEME` label is set
+  * `DCOS_SERVICE_PORT_INDEX` label is set
+  * task is in state `TASK_RUNNING`
+  If all the conditions above are met, request is routed using root Marathon
+  data. Please check the [section below](#recommended-way-to-expose-services)
+  for details on the meaning of these labels.
+* if the name of the service couldn't be resolved using root Marathon data, AR
+  scans the list of Mesos frameworks in search for the one with either framework
+  ID (higher priority) or framework name equal(lower priority) to
+  `namespace+service` string. If found, depending on the value of `webui_url` field:
+  * field contains an empty string - algorithm tries to resolve the
+    `namespace+service` name using MesosDNS
+  * request is proxied to the address pointed by the field contents
+  * field is `null` - resolving process for given `namespace and name` ends, MesosDNS
+    step is NOT executed
+* resolving via MesosDNS relies on issuing an HTTP SRV query to MesosDNS. For
+  `namespace and name` string equal to `foo/bar/baz`, the request that is going to
+  be made is:
+  ```
+  http://localhost:8123/baz.bar.foo._tcp.marathon.mesos
+  ```
+  In case when MesosDNS returns no results, AR assumes that service
+  `namespace and name` does run on DC/OS.
+
+#### Stale cache/broken cache cases
+`/service` endpoint internally uses AR cache, so an AR failing to refresh it
+has the same effects on it like any other endpoint. Please check the [Cache
+timers](#cache-timers) section for details.
+
+The thing worth remembering though, is that this endpoint uses more than one cache
+entry. Because of that, if e.g. root Marathon fails, and Mesos is fine, requests
```

Comment:
```
 373 cache entry. Because of that, if for example root Marathon fails, and Mesos is
~ 374 fine, requests will not be routed. This is because there has to be a strict
```